File: 3768.minimum-inversion-count-in-subarrays-of-fixed-length.py
#
# @lc app=leetcode id=3768 lang=python3
#
# [3768] Minimum Inversion Count in Subarrays of Fixed Length
#

# @lc code=start
class Solution:
    def minInversionCount(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if k == 1:
            return 0
        
        # Coordinate compression
        sorted_nums = sorted(set(nums))
        rank = {v: i + 1 for i, v in enumerate(sorted_nums)}
        max_rank = len(sorted_nums)
        
        # BIT for counting
        bit = [0] * (max_rank + 1)
        
        def update(i, delta=1):
            while i <= max_rank:
                bit[i] += delta
                i += i & (-i)
        
        def query(i):
            s = 0
            while i > 0:
                s += bit[i]
                i -= i & (-i)
            return s
        
        # Count inversions in first window
        inv_count = 0
        for i in range(k):
            r = rank[nums[i]]
            # Count elements > nums[i] already added
            inv_count += query(max_rank) - query(r)
            update(r)
        
        min_inv = inv_count
        
        # Slide the window
        for i in range(n - k):
            # Remove nums[i] from window
            r_old = rank[nums[i]]
            update(r_old, -1)
            removed_inv = query(r_old - 1)  # Elements < nums[i] remaining in window
            
            # Add nums[i+k] to window
            r_new = rank[nums[i + k]]
            added_inv = query(max_rank) - query(r_new)  # Elements > nums[i+k] in window
            update(r_new)
            
            inv_count = inv_count - removed_inv + added_inv
            min_inv = min(min_inv, inv_count)
        
        return min_inv
# @lc code=end

File: 3768.minimum-inversion-count-in-subarrays-of-fixed-length.cpp
//
// @lc app=leetcode id=3768 lang=cpp
//
// [3768] Minimum Inversion Count in Subarrays of Fixed Length
//

// @lc code=start
class Solution {
public:
    long long minInversionCount(vector<int>& nums, int k) {
        int n = nums.size();
        
        // Coordinate compression
        vector<int> sortedNums = nums;
        sort(sortedNums.begin(), sortedNums.end());
        sortedNums.erase(unique(sortedNums.begin(), sortedNums.end()), sortedNums.end());
        unordered_map<int, int> compress;
        for (int i = 0; i < (int)sortedNums.size(); i++) {
            compress[sortedNums[i]] = i + 1;
        }
        
        int m = sortedNums.size();
        vector<int> bit(m + 1, 0);
        
        auto update = [&](int i, int delta) {
            for (; i <= m; i += i & (-i))
                bit[i] += delta;
        };
        
        auto query = [&](int i) -> int {
            int sum = 0;
            for (; i > 0; i -= i & (-i))
                sum += bit[i];
            return sum;
        };
        
        // Count inversions in the first window
        long long inv = 0;
        for (int i = 0; i < k; i++) {
            int ci = compress[nums[i]];
            // Count elements > nums[i] that have been added
            inv += i - query(ci);
            update(ci, 1);
        }
        
        long long result = inv;
        
        // Slide the window
        for (int l = 1; l + k - 1 < n; l++) {
            int leftVal = compress[nums[l - 1]];
            int rightVal = compress[nums[l + k - 1]];
            
            // Remove nums[l-1] from BIT
            update(leftVal, -1);
            
            // Lost inversions: elements < nums[l-1] in the remaining window
            inv -= query(leftVal - 1);
            
            // Gained inversions: elements > nums[l+k-1] in the current window
            inv += (k - 1) - query(rightVal);
            
            // Add nums[l+k-1] to BIT
            update(rightVal, 1);
            
            result = min(result, inv);
        }
        
        return result;
    }
};
// @lc code=end